### PR TITLE
feat(rebalancer): track source-started actions in memory

### PR DESCRIPTION
**Pull request by Mo-Hussain**
Tracks source-started actions in memory without treating every exception as a broadcast. Typed proven pre-send failures release their action; attempted/acknowledged submissions retain source identity and reservation. Approval identity is tracked separately. Provider/receipt failures and source-committed failures cannot authorize duplicate work, including after the two-week intent TTL.

- Inventory movement, deposit and MCR callers record submission observations before waiting.
- Explorer recovery indexes message IDs and chain-qualified execution identities once per sync and updates those indexes as actions are recovered.
- Completed actions require a fresh balance poll before replanning; failed inventory reads fail the poll instead of supplying zero.
- Before-send and actual after-broadcast E2E hooks have distinct retry expectations.
- The simulator ingests confirmed Dispatch receipts directly and checks that every scheduled user transfer was observed, removing timing-dependent false success.

The action model remains process-local; reconcile transactions before any restart. Durable #9348 remains out of scope.

Validation: 628 unit tests at this stage, build/lint and simulator build pass. The identical final runtime passed all 46 E2Es and all 23 simulator scenarios. Parent: #9347; child: #9349.

Implementation head: `89cffacaa156dab67ee06d6c0aa46e08cfe5f4f7`. Fresh CI and the amd64 validation build must pass on the rewritten heads. [Provider validation coverage and operational gates](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/dcd094411799852aef123bae24babb09740aa133/typescript/rebalancer/docs/external-bridge-validation.md). This update made no production changes. This PR must remain draft until the operator explicitly asks to mark it ready.

### CODE REVIEW

File: typescript/rebalancer/src/core/InventoryRebalancer.ts [hunk 1]
```diff
@@ -1621,6 +1626,15 @@ export class InventoryRebalancer implements IInventoryRebalancer {
         privateKeys[sourceProtocol],
         `Missing inventory signer key for protocol ${sourceProtocol} (chain ${sourceChain})`,
       );
+
+      const action = await this.actionTracker.createRebalanceAction({
```
Comment (paulbalaji) [inline typescript/rebalancer/src/core/InventoryRebalancer.ts:1630]: [Medium] Distinguish proven pre-broadcast failures from ambiguous post-broadcast failures. This reservation has no execution identity until `execute()` fully returns. A validation/requote/API/approval/simulation error before any send leaves an unpollable `in_progress` action that suppresses retry for the process lifetime; the four existing inventory retry E2Es all fail this way. Conversely, if broadcast succeeds and confirmation/SDK execution times out, the same identity-less action cannot correlate with Explorer recovery, producing a duplicate recovered action or permanent stale movement. The reserve-before-send variants in `Rebalancer.ts:841` and `InventoryRebalancer.ts:1120` have the same window. Please record `txHash` at the broadcast boundary (callback/split send-confirm), fail/remove the reservation only when failure is definitively pre-send, and test both phases.

**Review by paulbalaji — COMMENTED** [review]
Reviewed the complete 15-file diff at exact head `049ed6da81302af895545f1b9aaef343ae6452e1` against true base `9dda85d0ef9539a72ab632b7032862d8d2f5af17`, including existing feedback (none), action-state invariants, all callers/variants, history, tests, changeset, and CI.

One actionable lifecycle issue is attached inline. This head is a stack rebase with the same stable patch ID as the reviewed predecessor. On that identical patch, all four inventory retry E2E shards and the aggregate check failed on the proven pre-broadcast half of this issue; current rebase checks are running.

File: typescript/rebalancer/src/e2e/inventory-erc20-weighted.e2e-test.ts [hunk 2]
```diff
@@ -407,110 +407,32 @@ describe('Erc20 Inventory WeightedStrategy E2E', function () {
       .withBalances('ERC20_INVENTORY_WEIGHTED_IMBALANCED')
       .build();
 
-    const initialBalances = await getErc20RouterBalances(
-      localProviders,
-      erc20DeployedAddresses,
-    );
-
-    // Cycle 1: Bridge fails — intent created but stays not_started, no actions
+    // A thrown execute call may have broadcast before losing its response.
```
Comment (paulbalaji) [inline typescript/rebalancer/src/e2e/inventory-erc20-weighted.e2e-test.ts:410]: [Medium] This injected failure is proven pre-broadcast, not ambiguous: `failNextExecute()` throws at the first line of `MockExternalBridge.execute`, before route parsing or any source send. Rewriting the test to expect process-lifetime suppression removes the retry regression instead of covering a post-broadcast timeout. Preserve the retry expectation here and add a separate fail-after-send hook/test; production needs a broadcast boundary so only truly ambiguous failures remain reserved.

File: typescript/rebalancer/src/tracking/ActionTracker.ts [hunk 3]
```diff
@@ -931,10 +923,24 @@ export class ActionTracker implements IActionTracker {
   }
 
   private async recoverAction(msg: ExplorerMessage): Promise<void> {
-    // Check if action already exists
-    const existing = await this.rebalanceActionStore.get(msg.msg_id);
+    const existingById = await this.rebalanceActionStore.get(msg.msg_id);
+    const existingByExecution = (await this.rebalanceActionStore.getAll()).find(
```
Comment (paulbalaji) [inline typescript/rebalancer/src/tracking/ActionTracker.ts:927]: [Low/perf] Avoid a full store scan for every recovered Explorer message. `initialize()`/`sync` already iterate every message, while `getAll()` allocates and `find()` scans all actions each time, making recovery O(messages x actions) and potentially quadratic as recovered actions are added. Build message-ID/normalized-tx-hash maps once per sync, or add indexed store lookups and update the index when execution identity is attached.

**Review by paulbalaji — COMMENTED** [review]
Rereviewed the complete current delta at exact head `e2fed875667575339d2211e412017159cb9f9295` against true base `33466d8f4f8603ab645dd9d67d64937a50efca19`, including the latest test-only amendment. The existing Medium thread remains current, unresolved, and unreplied: identity is reserved before `execute` but attached only after it returns, suppressing proven pre-broadcast failures while still lacking durable post-broadcast identity. The new tests codify that boundary error rather than model ambiguity; a distinct recovery-complexity issue is also inline. Production-focused tests (155) passed at the production-identical prior head; exact-head checks were still running at this snapshot.

**Review by paulbalaji — APPROVED** [review]
Reviewed own diff at `89cffacaa156dab67ee06d6c0aa46e08cfe5f4f7` and combined stack at `dcd094411799852aef123bae24babb09740aa133`.

The previous pre-send/post-send reservation and Explorer-recovery scan findings are addressed in the combined tree. Remaining optional performance work: group actions by intent once per pass so completed/partial-intent calculations avoid O(active intents x retained actions) scans; resolve independent monitor block-tag and supply reads with bounded concurrency while retaining the same finality tags and awaited cycle callback. These are follow-ups, not reasons to weaken reservations or parallelize signer sends. Type-audit nit: use typed WarpCore fixtures instead of the double assertions in `Monitor.test.ts`.

Latest combined-tree validation: dependency build and rebalancer lint passed; 658 rebalancer, 24 SDK, 12 Tron, 5 submission and 25 Helm tests passed (724 total). One unchanged LiFi test hit its live-network timeout, then passed focused and full-suite reruns. The previous combined head also passed 14 simulation scenario/loader tests and simulator build; the restack only changed the Tron options fallback plus its regression test. These are combined-tree checks, not per-layer or live-provider execution tests.

Current-head CI rollup: 20 cancelled, 6 failure, 50 skipped, 48 success. Successful reruns do not make this entire rollup green; resolve the remaining CI gate before merge.

GLM, Kimi, Muse and an independent adversarial pass were reconciled against source and focused reproductions. Code approval does not establish production execution readiness: documented configured provider-path fixtures, live execution and rollout evidence remain separate gates.

File: typescript/rebalancer/src/tracking/submission.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        ...(error.txHash ? { txHash: error.txHash } : {}),
+      });
+    } else if (submissionState === 'not_submitted') {
+      // An adapter without instrumented submission boundaries cannot prove that it did not send.
+      submissionState = 'unknown';
+      await tracker.updateRebalanceActionExecution(actionId, {
+        submissionState,
+      });
+    }
+    if (submissionState === 'not_submitted' && !pendingApproval)
```
Comment (paulbalaji) [inline typescript/rebalancer/src/tracking/submission.ts:81]: [P2] Add reconciliation for a definitively reverted primary transaction. A mined status=0 send retains submitted plus its hash here, but creates no Dispatch to recover; the tracker skips the message-less action and TTL never releases an in-progress reservation. A normal state/fee race after estimation can therefore suppress rebalancing for the process lifetime. With an available finalized reverted receipt, the actual tracker stayed in_progress through three syncs past TTL and never queried the receipt. Keep unknown outcomes reserved, but query known primary hashes and release/fail the action only after configured-finality proof that execution reverted. Cover finalized revert, pending and unavailable-receipt cases.

**Review by paulbalaji — COMMENTED** [review]
Reviewed own diff at `87d34ab9a3527c8d52a9bd62e2bd95b4bbc996bf` as part of stacks 9635 and 9636 (14 PRs).

The submission observers, action precreation and recovery indexes address the previously reported ambiguous-broadcast tracking and repeated-scan issues. A new reliability gap remains inline: finalized reverted primary transactions have no reconciliation path and keep reservations indefinitely. The actual-source probe ran action/intent sync three times past TTL with an available finalized status=0 receipt; neither reservation cleared and no receipt was queried. Preserve conservative handling for genuinely unknown or source-committed outcomes. Nonblocking test typing cleanup: Monitor.test.ts adds double casts to WarpCore; use typed fixtures or a narrower dependency interface.

Fresh combined-tree validation at `c7f454e83b201657f83bc733cb45f32b9983c0d9`: dependency builds and rebalancer lint passed; 736 rebalancer, 26 Helm, 24 SDK, 39 Tron and 5 submission tests passed (830 total). These are combined-tree checks, not independent tests of every intermediate head or live-provider/production execution.

Latest check-name view: 1 cancel, 38 pass, 29 skipping. agent-configs remains cancelled. Earlier red aggregate jobs sampled on this stack failed because prerequisite jobs were cancelled, not because their logs showed a test assertion failure; complete the cancelled validation on the current head. Independent Muse, GLM and Kimi review output was reconciled with source and prior threads; unverified model suggestions were excluded.
